src/core-layers/experimental/screen-grid-layer/screen-grid-layer.js

```diff
@@ -0,0 +1,256 @@
+      }
+    });
+    /* eslint-disable max-len */
+
+    const options = {
+      id: `${this.id}-aggregator`,
+      shaderCache: this.context.shaderCache
+    };
+    this.setState({
+      model: this._getModel(gl),
+      gpuGridAggregator: new GPUGridAggregator(gl, options)
+    });
+  }
+
+  shouldUpdateState({changeFlags}) {
+    return changeFlags.somethingChanged;
+  }
+
+  updateState({oldProps, props, changeFlags}) {
+    super.updateState({props, oldProps, changeFlags});
+
+    if (changeFlags.dataChanged) {
+      this._processData();
+    }
+
+    const cellSizeChanged = props.cellSizePixels !== oldProps.cellSizePixels;
+    if (cellSizeChanged || changeFlags.viewportChanged) {
+      this.updateAggregation();
+    }
+  }
+
+  draw({uniforms}) {
+    const {minColor, maxColor, parameters = {}} = this.props;
+    const {model, maxCount, cellScale} = this.state;
+    uniforms = Object.assign({}, uniforms, {minColor, maxColor, maxCount, cellScale});
+    model.draw({
+      uniforms,
+      parameters: Object.assign(
+        {
+          depthTest: false,
+          depthMask: false
+        },
+        parameters
+      )
+    });
+  }
+
+  _getModel(gl) {
+    return new Model(
+      gl,
+      Object.assign({}, this.getShaders(), {
+        id: this.props.id,
+        geometry: new Geometry({
+          drawMode: GL.TRIANGLE_FAN,
+          attributes: {
+            vertices: new Float32Array([0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0])
+          }
+        }),
+        isInstanced: true,
+        shaderCache: this.context.shaderCache
+      })
+    );
+  }
+
+  updateAggregation() {
+    const {width, height} = this.context.viewport;
+    const {cellSizePixels, data} = this.props;
+    const {gl} = this.context;
+
+    const MARGIN = 2;
+    const cellScale = new Float32Array([
+      (cellSizePixels - MARGIN) / width * 2,
+      -(cellSizePixels - MARGIN) / height * 2,
+      1
+    ]);
+    const numCol = Math.ceil(width / cellSizePixels);
+    const numRow = Math.ceil(height / cellSizePixels);
+    const numInstances = numCol * numRow;
+    const dataBytes = numInstances * 4 * 4;
+    const {positions, weights} = this.state;
+
+    const countsBuffer = new Buffer(gl, {
+      size: 4,
+      bytes: dataBytes,
+      type: GL.FLOAT,
+      instanced: 1
+    });
+
+    const aggregateData = this.state.gpuGridAggregator.run({
+      positions,
+      weights,
+      cellSize: [cellSizePixels, cellSizePixels],
+      viewport: this.context.viewport,
+      countsBuffer,
+      useGPU: false // TODO: this shouldn't be an option, remove
+    });
+
+    const {maxCount} = aggregateData;
+
+    this.setState({
+      cellScale,
+      numCol,
+      numRow,
+      numInstances,
+      maxCount
+    });
+
+    const attributeManager = this.getAttributeManager();
+
+    // TODO: AttributeManager should be able to just take new buffer for one or more attributes.
+    // data and numInstances shouldn't be needed.
```

Comment:
I think we may want an `attributeManager.setExternalBuffers` method? We don't want to force a full update, just register those buffers.